src/test_application/activescenes/common_scene.cpp

```diff
@@ -0,0 +1,83 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2022 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include "common_scene.h"
+
+#include <osp/Active/SysRender.h>
+
+#include <osp/Active/SysHierarchy.h>
+
+namespace testapp
+{
+
+CommonTestScene::~CommonTestScene()
+{
+    // Clear data first
+    for (on_cleanup_t *cleanup : m_onCleanup)
+    {
+        cleanup(*this);
+    }
+
+    osp::active::SysRender::clear_owners(m_drawing);
+    osp::active::SysRender::clear_resource_owners(m_drawingRes, *m_pResources);
+}
+
+void CommonTestScene::update_hierarchy_delete()
+{
+    using namespace osp::active;
+    // Cut deleted entities out of the hierarchy
+    SysHierarchy::update_delete_cut(
+            m_basic.m_hierarchy,
+            std::cbegin(m_delete), std::cend(m_delete));
+
+    // Create a new delete vector that includes the descendents of our current
+    // vector of entities to delete
+    m_deleteTotal = m_delete;
+    SysHierarchy::update_delete_descendents(
+            m_basic.m_hierarchy,
+            std::cbegin(m_delete), std::cend(m_delete),
+            [=] (ActiveEnt ent)
+    {
+        m_deleteTotal.push_back(ent);
+    });
+}
+
+void CommonTestScene::update_delete()
+{
+    auto first = std::cbegin(m_deleteTotal);
+    auto last = std::cend(m_deleteTotal);
+    osp::active::update_delete_basic                (m_basic,   first, last);
+    osp::active::SysRender::update_delete_drawing   (m_drawing, first, last);
+
+    // Delete entity IDs
+    for (ActiveEnt const ent : m_deleteTotal)
+    {
+        if (m_activeIds.exists(ent))
+        {
+            m_activeIds.remove(ent);
```

Comment:
doesnt .remove() have to do the same search logic as .exists() ?

shouldn't you just call remove() and let it be a no-op if the ent doesn't exist?

Reply:
IdRegistry::remove throws an exception because I didn't know any better. 

Removing an invalid ID should be considered an error most of the time; the user should never be keeping them around in the first place. I might have two versions of this function instead.

The only reason why a check is needed here, is because an entity can be deleted twice, which adds it twice to the m_deleteTotal vector. If m_deleteTotal is changed to a hierarchical bitset, then this wouldn't be possible

Reply:
If it is considered a "you did something really stupid", then an assert might be better than an exception.

But I suppose that doesn't really answer my concern about the double search.

What architectural challenges would you have to introduce the hierarchical bitset?

Reply:
Double search? It's just accessing a bitset, no searching involved.

"Architectural challenges" would mostly involve adding a function to allocate the bitset, as well as changing a push_back(...) to a set(...).